// node-google-translator/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Translator = void 0;
exports.translate = translate;
const node_fetch_1 = __importDefault(require("node-fetch"));
const http_errors_1 = __importDefault(require("http-errors"));
const helpers_js_1 = require("./helpers.js");
const defaults = {
    from: 'auto',
    to: 'en',
    host: 'translate.google.com',
};
class Translator {
    constructor(inputText, options) {
        this.inputText = inputText;
        this.options = Object.assign({}, defaults, options);
    }
    async translate() {
        const url = this.buildUrl();
        const fetchOptions = this.buildFetchOptions();
        const res = await (0, node_fetch_1.default)(url, fetchOptions);
        if (!res.ok) {
            throw await this.buildError(res);
        }
        const raw = await res.json();
        const text = this.buildResText(raw);
        return { text, raw };
    }
    buildUrl() {
        const { host } = this.options;
        return [
            `https://${host}/translate_a/single`,
            '?client=at',
            '&dt=t', // return sentences
            '&dt=rm', // add translit to sentences
            '&dj=1', // result as pretty json instead of deep nested arrays
        ].join('');
    }
    buildBody() {
        const { from, to } = this.options;
        const params = new URLSearchParams();
        params.append('sl', from);
        params.append('tl', to);
        params.append('q', this.inputText);
        return params;
    }
    buildFetchOptions() {
        const { fetchOptions } = this.options;
        const res = Object.assign({}, fetchOptions);
        res.method = 'POST';
        res.headers = Object.assign({}, res.headers, {
            'Content-Type': 'application/x-www-form-urlencoded;charset=utf-8',
        });
        res.body = this.buildBody();
        return res;
    }
    buildResText({ sentences }) {
        return sentences
            .filter((s) => 'trans' in s)
            .map(s => s.trans)
            .join('');
    }
    async buildError(res) {
        if (res.status === 429) {
            const text = await res.text();
            const { ip, time, url } = (0, helpers_js_1.extractTooManyRequestsInfo)(text);
            const message = `${res.statusText} IP: ${ip}, Time: ${time}, Url: ${url}`;
            return (0, http_errors_1.default)(res.status, message);
        }
        else {
            return (0, http_errors_1.default)(res.status, res.statusText);
        }
    }
}
exports.Translator = Translator;
async function translate(inputText, options) {
    return new Translator(inputText, options).translate();
}
__exportStar(require("./types.js"), exports);
